cel/validator.go

```diff
@@ -0,0 +1,282 @@
+func (homogeneousAggregateLiteralValidator) Name() string {
+	return "cel.lib.std.types.homogeneous"
+}
+
+// Validate validates that all lists and map literals have homogeneous types, i.e. don't contain dyn types.
+//
+// This validator makes an exception for list and map literals which occur at any level of nesting within
+// string format calls.
+func (v homogeneousAggregateLiteralValidator) Validate(_ *Env, a *Ast, iss *Issues) {
+	errs := errorReporter{iss: iss, info: a.info}
+	root := ast.NavigateCheckedAST(astToCheckedAST(a))
+	listExprs := ast.MatchDescendants(root, ast.KindMatcher(ast.ListKind))
+	for _, listExpr := range listExprs {
+		if hasStringFormatAncestor(listExpr) {
+			continue
+		}
+		l := listExpr.AsList()
+		elements := l.Elements()
+		optIndices := l.OptionalIndices()
+		var elemType *Type
+		for i, e := range elements {
+			et := e.Type()
+			if isOptionalIndex(i, optIndices) {
+				et = et.Parameters()[0]
+			}
+			if elemType == nil {
+				elemType = et
+				continue
+			}
+			if !elemType.IsEquivalentType(et) {
+				v.typeMismatch(errs, e.ID(), elemType, et)
+				break
+			}
+		}
+	}
+	mapExprs := ast.MatchDescendants(root, ast.KindMatcher(ast.MapKind))
+	for _, mapExpr := range mapExprs {
+		if hasStringFormatAncestor(mapExpr) {
+			continue
+		}
+		m := mapExpr.AsMap()
+		entries := m.Entries()
+		var keyType, valType *Type
+		for _, e := range entries {
+			key, val := e.Key(), e.Value()
+			kt, vt := key.Type(), val.Type()
+			if e.IsOptional() {
+				vt = vt.Parameters()[0]
+			}
+			if keyType == nil && valType == nil {
+				keyType, valType = kt, vt
+				continue
+			}
+			if !keyType.IsEquivalentType(kt) {
+				v.typeMismatch(errs, key.ID(), keyType, kt)
+			}
+			if !valType.IsEquivalentType(vt) {
+				v.typeMismatch(errs, val.ID(), valType, vt)
+			}
+		}
+	}
+}
+
+func hasStringFormatAncestor(e ast.NavigableExpr) bool {
+	if parent, found := e.Parent(); found {
+		if parent.Kind() == ast.CallKind && parent.AsCall().FunctionName() == "format" {
+			return true
+		}
+		if parent.Kind() == ast.ListKind || parent.Kind() == ast.MapKind {
+			return hasStringFormatAncestor(parent)
+		}
+	}
+	return false
+}
+
+func isOptionalIndex(i int, optIndices []int32) bool {
+	for _, optInd := range optIndices {
+		if i == int(optInd) {
+			return true
+		}
+	}
+	return false
+}
+
+func (homogeneousAggregateLiteralValidator) typeMismatch(errs errorReporter, id int64, expected, actual *Type) {
+	errs.reportErrorAtID(id, "expected type '%s' but found '%s'", FormatCelType(expected), FormatCelType(actual))
+}
+
+func (homogeneousAggregateLiteralValidator) is_validator() {}
+
+type errorReporter struct {
+	iss  *Issues
+	info *exprpb.SourceInfo
+}
+
+func (er *errorReporter) reportErrorAtID(id int64, message string, args ...any) {
+	er.iss.errs.ReportErrorAtID(id, locationByID(id, er.info), message, args...)
+}
+
+func locationByID(id int64, sourceInfo *exprpb.SourceInfo) common.Location {
```

Comment:
Some of the functions here, such as this seems useful enough to offer in a common helper.

Reply:
I ended up moving most of this functionality up into the `Issues` object in a later PR. See https://github.com/google/cel-go/pull/768 cel/env.go line 693. I can probably move that change into its own PR if you want.

Reply:
This is OK, we can bring in the change with the linked PR